File: app/web/blog/services/comment_service.py
from datetime import datetime, timedelta
from typing import List, Optional, Tuple, Dict

from flask_login import current_user
from markupsafe import escape

from app.extensions import db
from app.models import Blog, BlogComment
from app.service.notifications import send_notification


class CommentService:
    """评论业务逻辑服务"""

    @staticmethod
    def _serialize_comment(comment: BlogComment) -> Dict:
        if comment.is_deleted:
            content_html = "[该评论已删除]"
        else:
            content_html = comment.content_html or ''
        return {
            'id': comment.id,
            'blog_id': comment.blog_id,
            'author': {
                'id': comment.author.id if comment.author else None,
                'username': comment.author.username if comment.author else None,
                'is_admin': getattr(comment.author, 'has_admin_rights', False) if comment.author else False,
            },
            'parent_id': comment.parent_id,
            'root_id': comment.root_id,
            'content_html': content_html,
            'status': comment.status,
            'is_deleted': comment.is_deleted,
            'likes_count': comment.likes_count,
            'created_at': comment.created_at.isoformat() if comment.created_at else None,
            'updated_at': comment.updated_at.isoformat() if comment.updated_at else None,
            'children': []  # 由构建树方法填充
        }

    @staticmethod
    def _build_comment_tree(comments: List[BlogComment]) -> List[Dict]:
        """将评论列表构建为嵌套树结构。"""
        id_to_node: Dict[str, Dict] = {}
        roots: List[Dict] = []

        # 初始化节点
        for c in comments:
            id_to_node[c.id] = CommentService._serialize_comment(c)

        # 建树
        for c in comments:
            node = id_to_node[c.id]
            if c.parent_id and c.parent_id in id_to_node:
                parent_node = id_to_node[c.parent_id]
                parent_node['children'].append(node)
            else:
                roots.append(node)

        # 递归对每个 children 按时间排序
        def sort_children(n: Dict):
            n['children'].sort(key=lambda x: x['created_at'] or '')
            for ch in n['children']:
                sort_children(ch)

        for r in roots:
            sort_children(r)

        # 顶层按时间排序
        roots.sort(key=lambda x: x['created_at'] or '')
        return roots

    @staticmethod
    def list_comments(blog_id: str) -> List[Dict]:
        """获取某篇文章的所有评论（已批准且未删除），返回树结构。"""
        comments = BlogComment.query.filter_by(
            blog_id=blog_id,
            status='approved'
            #is_deleted=False
        ).order_by(BlogComment.created_at.asc()).all()

        if not comments:
            return []

        return CommentService._build_comment_tree(comments)

    @staticmethod
    def create_comment(blog_id: str, content: str, parent_id: Optional[str] = None) -> Tuple[bool, str, Optional[Dict]]:
        """
        创建评论。

        规则：
        - 仅核心用户或管理员可发评论
        - 禁言用户禁止发言（管理员不受限）
        - 频率限制：30 秒 1 条
        - 内容长度：1..2000
        - 不支持 Markdown：保存时进行 HTML 转义，并将换行转 <br>
        - 楼中楼：parent_id 可选；root_id 指向最顶层评论
        """
        # 权限校验
        if not getattr(current_user, 'is_core_user', False):
            return False, '只有核心用户或管理员可以发表评论', None

        # 禁言检查（管理员除外）
        if not getattr(current_user, 'has_admin_rights', False) and current_user.is_currently_banned():
            return False, '您已被禁言，无法发表评论', None

        # 频率限制：最近 30 秒内是否已有评论
        one_minute_ago = datetime.now() - timedelta(seconds=30)
        recent_exists = BlogComment.query.filter(
            BlogComment.author_id == current_user.id,
            BlogComment.created_at >= one_minute_ago,
            BlogComment.is_deleted == False
        ).first()
        if recent_exists:
            return False, '发言过于频繁，请稍后再试', None

        # 验证文章存在
        blog = Blog.query.get(blog_id)
        if not blog or blog.ignore:
            return False, '文章不存在', None

        # 验证内容
        if content is None:
            return False, '评论内容不能为空', None
        content = content.strip()
        if not content:
            return False, '评论内容不能为空', None
        if len(content) > 2000:
            return False, '评论内容不能超过2000字', None

        # 处理父评论与 root_id
        parent = None
        root_id: Optional[str] = None
        if parent_id:
            parent = BlogComment.query.get(parent_id)
            if not parent or parent.blog_id != blog_id or parent.is_deleted:
                return False, '父评论不存在或已删除', None
            root_id = parent.root_id or parent.id

        # 构建 content_html：仅转义 + 换行转 <br>
        escaped = escape(content)
        content_html = str(escaped).replace('\n', '<br>')

        # 创建记录
        comment = BlogComment(
            blog_id=blog_id,
            author_id=current_user.id,
            parent_id=parent.id if parent else None,
            root_id=root_id,
            content=content,
            content_html=content_html,
            status='approved',
            is_deleted=False,
            created_at=datetime.now(),
        )
        db.session.add(comment)

        # 维护博客冗余字段
        blog.comments_count = (blog.comments_count or 0) + 1
        blog.last_comment_at = datetime.now()

        db.session.commit()

        # 发送通知：回复则通知被回复者；否则通知文章作者（排除自己）
        try:
            if parent and parent.author_id != current_user.id:
                send_notification(
                    recipient_id=parent.author_id,
                    action='评论回复',
                    actor_id=current_user.id,
                    object_type='blog',
                    object_id=blog_id,
                    detail=f'你的评论在《{blog.title}》下收到了回复'
                )
            elif not parent and blog.author_id != current_user.id:
                send_notification(
                    recipient_id=blog.author_id,
                    action='文章评论',
                    actor_id=current_user.id,
                    object_type='blog',
                    object_id=blog_id,
                    detail=f'你的文章《{blog.title}》收到了新评论'
                )
        except Exception:
            # 通知失败不影响主流程
            pass

        return True, '评论成功', CommentService._serialize_comment(comment)

    @staticmethod
    def delete_comment(comment_id: str, reason: Optional[str] = None) -> Tuple[bool, str]:
        """删除（软删）评论：作者本人或管理员可删除。
        - 管理员删除他人评论时必须给出删除原因（1..500）。
        """
        comment = BlogComment.query.get(comment_id)
        if not comment or comment.is_deleted:
            return False, '评论不存在或已删除'

        is_admin = getattr(current_user, 'has_admin_rights', False)
        is_owner = (comment.author_id == current_user.id)

        if not (is_admin or is_owner):
            return False, '无权删除该评论'

        admin_deleting_others = is_admin and not is_owner
        if admin_deleting_others:
            reason = (reason or '').strip()
            if not reason:
                return False, '请提供删除原因'
            if len(reason) > 500:
                return False, '删除原因过长（最多500字）'

        comment.is_deleted = True
        # 可选：清空展示内容，保留审计
        # comment.content = ''
        # comment.content_html = ''

        # 维护计数
        blog = Blog.query.get(comment.blog_id)
        if blog and (blog.comments_count or 0) > 0:
            blog.comments_count = (blog.comments_count or 0) - 1

        was_admin_delete = admin_deleting_others
        db.session.commit()

        # 记录管理员操作日志（管理员删除他人评论）
        if was_admin_delete:
            try:
                from app.service.audit_log import log_admin_action
                log_admin_action(
                    action='delete_comment',
                    admin_id=current_user.id,
                    target_user_id=comment.author_id,
                    object_type='comment',
                    object_id=comment.id,
                    reason=reason or '违反规则',
                    metadata={'blog_id': comment.blog_id}
                )
            except Exception:
                pass
        return True, '删除成功'

File: web-next/src/lib/comment-service.ts
// ─────────────────────────────────────────────────────────────────────────────
// comment-service.ts — 评论业务逻辑（对齐 Flask CommentService）
//
// 楼中楼：parentId（直接父级）+ rootId（顶层评论串）。评论不渲染 Markdown：
// 保存时仅 HTML 转义并把换行转 <br> 写入 contentHtml（对齐 markupsafe.escape）。
// 软删除：isDeleted=true。列表时构建树并丢弃「无子评论的已删除叶子」
// （对齐 _filter_deleted_leaves）。计数在事务内按「未删除数」重算。
// ─────────────────────────────────────────────────────────────────────────────

import { prisma } from './db';
import { nowForDb } from './db-time';
import { hasAdminRights } from './auth';
import { rateLimit, RULES } from './rate-limit';
import { sendNotification } from './notification-service';
import { logAdminAction } from './admin-user-service';
import type { Prisma } from '@prisma/client';

// ── 序列化输出（snake_case，对齐 Flask API JSON 形状）──────────────────────────

export interface CommentNode {
  id: string;
  blog_id: string;
  author: {
    id: string | null;
    username: string | null;
    is_admin: boolean;
    avatar_url: string | null;
  };
  parent_id: string | null;
  root_id: string | null;
  content_html: string;
  status: string | null;
  is_deleted: boolean;
  likes_count: number;
  created_at: string | null;
  updated_at: string | null;
  children: CommentNode[];
}

const DELETED_PLACEHOLDER = '[该评论已删除]';

// markupsafe.escape 语义：& < > " ' → 实体
function escapeHtml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&#34;')
    .replace(/'/g, '&#39;');
}

/** 评论内容 → contentHtml：仅转义 + 换行转 <br>（不支持 Markdown）。 */
export function toContentHtml(content: string): string {
  return escapeHtml(content).replace(/\n/g, '<br>');
}

// Prisma 行的最小选择集（含作者用于序列化）
const commentSelect = {
  id: true,
  blogId: true,
  authorId: true,
  parentId: true,
  rootId: true,
  contentHtml: true,
  status: true,
  isDeleted: true,
  likesCount: true,
  createdAt: true,
  updatedAt: true,
  author: { select: { id: true, username: true, role: true } },
} satisfies Prisma.BlogCommentSelect;

type CommentRow = Prisma.BlogCommentGetPayload<{ select: typeof commentSelect }>;

function serializeRow(c: CommentRow): CommentNode {
  const deleted = c.isDeleted ?? false;
  return {
    id: c.id,
    blog_id: c.blogId,
    author: {
      id: c.author?.id ?? null,
      username: c.author?.username ?? null,
      is_admin: c.author ? hasAdminRights(c.author) : false,
      avatar_url: c.author ? `/api/avatar/${c.author.id}` : null,
    },
    parent_id: c.parentId,
    root_id: c.rootId,
    content_html: deleted ? DELETED_PLACEHOLDER : c.contentHtml ?? '',
    status: c.status,
    is_deleted: deleted,
    likes_count: c.likesCount ?? 0,
    created_at: c.createdAt ? c.createdAt.toISOString() : null,
    updated_at: c.updatedAt ? c.updatedAt.toISOString() : null,
    children: [],
  };
}

/** 递归移除「无子评论的已删除评论」（对齐 _filter_deleted_leaves）。 */
function filterDeletedLeaves(nodes: CommentNode[]): CommentNode[] {
  const result: CommentNode[] = [];
  for (const node of nodes) {
    node.children = filterDeletedLeaves(node.children);
    if (node.is_deleted && node.children.length === 0) continue;
    result.push(node);
  }
  return result;
}

/**
 * 获取某文章的评论树（status='approved'，含已删除节点参与建树，
 * 最后丢弃无子的已删除叶子）。输入已按 createdAt 升序，天然保序。
 */
export async function listCommentsForBlog(blogId: string): Promise<CommentNode[]> {
  const rows = await prisma.blogComment.findMany({
    where: { blogId, status: 'approved' },
    orderBy: { createdAt: 'asc' },
    select: commentSelect,
  });
  if (rows.length === 0) return [];

  const idToNode = new Map<string, CommentNode>();
  for (const r of rows) idToNode.set(r.id, serializeRow(r));

  const roots: CommentNode[] = [];
  for (const r of rows) {
    const node = idToNode.get(r.id)!;
    const parent = r.parentId ? idToNode.get(r.parentId) : undefined;
    if (parent) parent.children.push(node);
    else roots.push(node);
  }

  return filterDeletedLeaves(roots);
}

// ── 创建 ─────────────────────────────────────────────────────────────────────

export interface CreateCommentInput {
  blogId: string;
  authorId: string;
  content: string;
  parentId?: string | null;
}

export type CreateCommentResult =
  | { ok: true; comment: CommentNode }
  | {
      ok: false;
      error: 'rateLimited' | 'notFound' | 'empty' | 'tooLong' | 'parentInvalid';
      message: string;
    };

/**
 * 创建评论（对齐 CommentService.create_comment）。
 * 调用方负责登录 / 禁言校验；此处负责频率限制、内容校验、建 root_id、维护冗余计数。
 */
export async function createComment(input: CreateCommentInput): Promise<CreateCommentResult> {
  const { blogId, authorId, parentId } = input;

  // 每日频率限制（对齐 RULES.commentDaily，按用户键）
  const daily = rateLimit(`comment:d:${authorId}`, RULES.commentDaily);
  if (!daily.allowed) {
    return { ok: false, error: 'rateLimited', message: '今日评论已达上限（1200条），请明日再试' };
  }

  const content = (input.content ?? '').trim();
  if (!content) return { ok: false, error: 'empty', message: '评论内容不能为空' };
  if (content.length > 2000) return { ok: false, error: 'tooLong', message: '评论内容不能超过2000字' };

  const contentHtml = toContentHtml(content);
  const now = nowForDb();
  const id = crypto.randomUUID();

  try {
    const node = await prisma.$transaction(async (tx) => {
      // 带出 title/authorId 供事务提交后发通知（对齐 Flask 的评论通知）
      const blog = await tx.blog.findFirst({
        where: { id: blogId, ignore: false },
        select: { id: true, title: true, authorId: true },
      });
      if (!blog) return { notFound: true as const };

      let resolvedParentId: string | null = null;
      let rootId: string | null = null;
      let parentAuthorId: string | null = null;
      if (parentId) {
        const parent = await tx.blogComment.findUnique({
          where: { id: parentId },
          select: { id: true, blogId: true, rootId: true, isDeleted: true, authorId: true },
        });
        if (!parent || parent.blogId !== blogId || parent.isDeleted) {
          return { parentInvalid: true as const };
        }
        resolvedParentId = parent.id;
        rootId = parent.rootId ?? parent.id;
        parentAuthorId = parent.authorId;
      }

      const created = await tx.blogComment.create({
        data: {
          id,
          blogId,
          authorId,
          parentId: resolvedParentId,
          rootId,
          content,
          contentHtml,
          status: 'approved',
          isDeleted: false,
          likesCount: 0,
          createdAt: now,
          updatedAt: now,
        },
        select: commentSelect,
      });

      const commentsCount = await tx.blogComment.count({ where: { blogId, isDeleted: false } });
      await tx.blog.update({ where: { id: blogId }, data: { commentsCount, lastCommentAt: now } });

      return {
        row: created,
        notify: {
          blogTitle: blog.title,
          blogAuthorId: blog.authorId,
          parentAuthorId,
        },
      };
    });

    if ('notFound' in node) return { ok: false, error: 'notFound', message: '文章不存在' };
    if ('parentInvalid' in node) return { ok: false, error: 'parentInvalid', message: '父评论不存在或已删除' };

    // 发送通知（对齐 Flask CommentService.create_comment）：
    //   回复 → 通知被回复者；顶层 → 通知文章作者；两者都排除「自己评自己」。
    // 与 Flask 一致：通知失败不影响主流程（评论已提交成功），故整体 try/catch 吞掉。
    try {
      const { blogTitle, blogAuthorId, parentAuthorId: pAuthor } = node.notify;
      if (pAuthor && pAuthor !== authorId) {
        await sendNotification({
          recipientId: pAuthor,
          action: '评论回复',
          actorId: authorId,
          objectType: 'blog',
          objectId: blogId,
          detail: `你的评论在《${blogTitle}》下收到了回复`,
        });
      } else if (!pAuthor && blogAuthorId && blogAuthorId !== authorId) {
        await sendNotification({
          recipientId: blogAuthorId,
          action: '文章评论',
          actorId: authorId,
          objectType: 'blog',
          objectId: blogId,
          detail: `你的文章《${blogTitle}》收到了新评论`,
        });
      }
    } catch {
      // 通知失败不影响评论本身（对齐 Flask 的 try/except pass）
    }

    return { ok: true, comment: serializeRow(node.row) };
  } catch {
    return { ok: false, error: 'notFound', message: '文章不存在' };
  }
}

// ── 软删除 ───────────────────────────────────────────────────────────────────

export type DeleteActor = { id: string; role: string };

export type DeleteCommentResult =
  | { ok: true }
  | {
      ok: false;
      // reasonRequired / reasonTooLong：管理员删他人评论时的原因校验（对齐 Flask）
      error: 'notFound' | 'forbidden' | 'reasonRequired' | 'reasonTooLong';
      message: string;
    };

/**
 * 软删除评论（对齐 delete_comment）：作者本人或管理员可删。
 * 删除后重算文章未删除评论数与 lastCommentAt。
 *
 * @param reason 管理员删「他人」评论时必填（1..500）；作者删自己的可省略。
 *               该原因会写入 AdminActionLog —— /audit 公示与用户申诉依赖它。
 */
export async function softDeleteComment(
  commentId: string,
  actor: DeleteActor,
  reason?: string
): Promise<DeleteCommentResult> {
  const outcome = await prisma.$transaction(async (tx) => {
    const comment = await tx.blogComment.findUnique({
      where: { id: commentId },
      select: { id: true, blogId: true, authorId: true, isDeleted: true },
    });
    if (!comment || comment.isDeleted) {
      return { ok: false as const, error: 'notFound' as const, message: '评论不存在或已删除' };
    }

    const isAuthor = comment.authorId === actor.id;
    if (!isAuthor && !hasAdminRights(actor)) {
      return { ok: false as const, error: 'forbidden' as const, message: '无权删除该评论' };
    }

    // 对齐 Flask：管理员删「他人」评论时必须给出原因（1..500），作者删自己的不需要。
    const adminDeletingOthers = !isAuthor && hasAdminRights(actor);
    const trimmedReason = (reason ?? '').trim();
    if (adminDeletingOthers) {
      if (!trimmedReason) {
        return { ok: false as const, error: 'reasonRequired' as const, message: '请提供删除原因' };
      }
      if (trimmedReason.length > 500) {
        return { ok: false as const, error: 'reasonTooLong' as const, message: '删除原因过长（最多500字）' };
      }
    }

    await tx.blogComment.update({ where: { id: commentId }, data: { isDeleted: true } });

    const commentsCount = await tx.blogComment.count({ where: { blogId: comment.blogId, isDeleted: false } });
    const latest = await tx.blogComment.findFirst({
      where: { blogId: comment.blogId, isDeleted: false },
      orderBy: { createdAt: 'desc' },
      select: { createdAt: true },
    });
    await tx.blog.update({
      where: { id: comment.blogId },
      data: { commentsCount, lastCommentAt: latest?.createdAt ?? null },
    });

    return {
      ok: true as const,
      audit: adminDeletingOthers
        ? { targetUserId: comment.authorId, blogId: comment.blogId, reason: trimmedReason }
        : null,
    };
  });

  // 记录管理员操作日志（对齐 Flask：管理员删他人评论才记）。
  // 这条日志是 /audit 公示与申诉流程的数据来源 —— 缺了用户就无法申诉。
  // 与 Flask 一致：日志失败不回滚删除本身。
  if (outcome.ok && outcome.audit) {
    try {
      await logAdminAction({
        action: 'delete_comment',
        adminId: actor.id,
        targetUserId: outcome.audit.targetUserId,
        objectType: 'comment',
        objectId: commentId,
        reason: outcome.audit.reason || '违反规则',
        metadata: { blog_id: outcome.audit.blogId },
      });
    } catch {
      /* 审计写入失败不影响删除结果（对齐 Flask 的 try/except pass） */
    }
  }

  if (outcome.ok) return { ok: true };
  // 剥掉内部用的 audit 字段，只暴露对外契约
  const { ok, error, message } = outcome;
  return { ok, error, message };
}

// ── 点赞切换 ─────────────────────────────────────────────────────────────────

export type ToggleLikeResult =
  | { liked: boolean; likesCount: number }
  | { notFound: true };

/**
 * 切换评论点赞（唯一约束 commentId+userId），维护 BlogComment.likesCount。
 */
export async function toggleCommentLike(commentId: string, userId: string): Promise<ToggleLikeResult> {
  return prisma.$transaction(async (tx) => {
    const comment = await tx.blogComment.findUnique({
      where: { id: commentId },
      select: { id: true, isDeleted: true },
    });
    if (!comment || comment.isDeleted) return { notFound: true as const };

    const existing = await tx.commentLike.findUnique({
      where: { uq_comment_like_comment_user: { commentId, userId } },
    });

    let liked: boolean;
    if (existing) {
      await tx.commentLike.delete({ where: { id: existing.id } });
      liked = false;
    } else {
      await tx.commentLike.create({ data: { commentId, userId, createdAt: nowForDb() } });
      liked = true;
    }

    const likesCount = await tx.commentLike.count({ where: { commentId } });
    await tx.blogComment.update({ where: { id: commentId }, data: { likesCount } });
    return { liked, likesCount };
  });
}
